### ISSUE TITLE
Problem: There is no API or mechanism to manage ROUTER socket connected clients

### ISSUE THREAD

**Issue by reza-ebrahimi** [zeromq/libzmq#2319]
There is 10 Clients (DEALER socket) that connected to a ROUTER socket, if all of clients send message to ROUTER socket and in ROUTER side for sending reply response to exactly such DEALER it might know about its handle, to do that there is no API to manage DEALER handles in ROUTER side, so manager should do some tasks like adding, removing or query a handle.

**Comment by bluca** [comment]
Isn't the identity used for that purpose?

**Comment by reza-ebrahimi** [comment]
@bluca  Yes, of course! but identity doesn't enough in dealing with some situations, for example, one problem happened when a DEALER socket lost (disconnected) its connection, so that socket identity still be in ROUTER socket side and exists their identity state in application code, if you send any data to disconnected DEALER socket it will face with access violation error. so in cause of the error, application code doesn't know about disconnected socket state or doesn't notified about that to consider another tasks such as removing socket identity handle.

I think this is a ZeroMQ kernel task to provide a identity manager in ROUTER socket side and do proper tasks (adding, removing, querying), by now there is no notifying mechanism from ROUTER socket about disconnection of ROUTER clients.

**Comment by reza-ebrahimi** [comment]
@bluca have you any idea?

**Comment by bluca** [comment]
Isn't ROUTER_MANDATORY https://github.com/zeromq/libzmq/blob/master/doc/zmq_setsockopt.txt#L705 a solution for that already? It will return an error in such cases

**Comment by mrvn** [comment]
The socket monitor will send a disconnect event and the application can remove the DEALER. Isn't that all you need?

**Comment by reza-ebrahimi** [comment]
Yes, That is! I need to know which peer disconnected, to do this zeromq might provide me at least peer's identity and another useful information such as socket lost reason:

Disconnection event structure:
  - socket identity
  - disconnection reason

**Comment by bluca** [comment]
PRs are most welcome :)

**Comment by reza-ebrahimi** [comment]
If we want to modify zmq monitoring system, we should use ZMQ_PAIR socket, 
documentation says:
> A socket of type ZMQ_PAIR can only be connected to a single peer at any one time. No message routing or filtering is performed on messages sent over a ZMQ_PAIR socket.

in this case we should create and connect multiple instance of ZMQ_PAIR socket based on how many peers connected to ROUTER socket, I think this isn't efficient, any idea are welcome!

**Comment by nishsharma001** [comment]
Guys, 
           This is requirement that we were looking for in our application as well. Has this been addressed in any form ? 
Please guide.

**Comment by bluca** [comment]
It has not - pull requests are welcome

**Comment by reza-ebrahimi** [comment]
@bluca I'll be work on it by next week.

**Comment by nagarajsherigar** [comment]
Hi
We also wanted a solution like this. Is it available now?

**Comment by dbivolaru** [comment]
A disconnection is notified using [identity, b''] (ie one-message-with-identity-and-empty-string) in pyzmq so you actually know who and when it disconnected. I would assume it's similar in the lower level api.

**Comment by chinmayprabhudesai** [comment]
@dbivolaru ,

That only happens if you set appropriate flags for `ZMQ_ROUTER_NOTIFY` socketopt.

From [libzmq/master docs](http://api.zeromq.org/master:zmq-setsockopt#toc81), it seems that this option is still in draft state and not part of the general release.

Did you build libzmq  with `--enable-drafts` option?

If not, is this socketopt out of it's draft state?

Thanks!

**Comment by getarobo** [comment]
Any updates or approachs on this? My server with router keep crashing when I try to send msg to  disconnected dealer(due to bad internet).    NEED HELP !

**Comment by getarobo** [comment]
> Any updates or approachs on this? My server with router keep crashing when I try to send msg to disconnected dealer(due to bad internet). NEED HELP !

NVM  setting zmq_router_manadatory to 0 did the trick which is default.
```
zmq::socket_t cms_router(*ctx, zmq::socket_type::router);
cms_router.set(zmq::sockopt::router_mandatory,0); //<--- ignore error when dealer doesnt exist
cms_router.bind("tcp://*:" + port);
```
